### PR TITLE
feat: allow for configuration of the async writers executors

### LINKED ISSUE #587: Allow for configuration of the async writers' executor(s)
### Persona

As a Block Node Operator/Developer

### Request

I want to have configurable values for the thread pool executor that is inside the async writer

### Goal

So that I can have control over the executor's internal properties when it is being created

### Technical Notes

- configurable values need to exist so that the executor (the one that is being submitted async persistence tasks) that is internal to the persistence

### PR DESCRIPTION
**Pull request by georgi-l95**
This PR adds a configurable executor for async writers in the persistence storage module. The implementation allows for different executor types and configurations to optimize performance for various workloads and deployment environments.
#### Key Changes:
- Added `ExecutorType` enum to `PersistenceStorageConfig` with options for `THREAD_POOL`, `SINGLE_THREAD`, and `FORK_JOIN` executors
- Added configuration properties for thread count, thread keep-alive time, virtual threads support, and execution queue limit
- Implemented `AsyncWriterExecutorFactory` to create appropriate executor instances based on configuration
- Updated `PersistenceInjectionModule` to use the factory for creating executors
- Added unit tests for all new components

#### Configuration Options:
- `persistence.storage.executionQueueLimit`: Maximum queue size for pending tasks (64-2048)
- `persistence.storage.executorType`: Type of executor (THREAD_POOL, SINGLE_THREAD, FORK_JOIN)
- `persistence.storage.threadCount`: Number of threads for thread pool (1-16)
- `persistence.storage.threadKeepAliveTime`: Keep-alive time in seconds for idle threads
- `persistence.storage.useVirtualThreads`: Whether to use virtual threads (boolean)
## Related Issue(s)
Fixes #587

### CODE REVIEW

File: server/src/main/java/com/hedera/block/server/persistence/storage/PersistenceStorageConfig.java [hunk 1]
```diff
@@ -66,6 +71,33 @@ public enum StorageType {
...
+        THREAD_POOL,
+
+        /**
+         * An executor with a single worker thread.
+         * Suitable for low-throughput environments or when strict ordering of task execution is required.
+         */
+        SINGLE_THREAD,
+
+        /**
+         * Uses the common ForkJoinPool for task execution.
```
Comment (jsync-swirlds) [inline server/src/main/java/com/hedera/block/server/persistence/storage/PersistenceStorageConfig.java:94]: This should use _a_ fork/join pool, but should not use the _common_ fork/join pool.
The reason for this is that the common pool is almost always overused and in container environments often ends up having much fewer threads than it should.
In my opinion, we should, if this is selected, create a separate fork/join pool with a thread count set by the threadCount config value.

File: server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+     *
+     * @return the common fork-join pool
+     */
+    @NonNull
+    private static ForkJoinPool createForkJoinExecutor() {
+        final ForkJoinPool forkJoinPool = ForkJoinPool.commonPool();
+        LOGGER.log(
+                TRACE,
+                "Using common fork-join pool with parallelism level: %d".formatted(forkJoinPool.getParallelism()));
+        return forkJoinPool;
```
Comment (jsync-swirlds) [inline server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java:134]: We should not use the common pool here. We should treat this very much like the thread pool case, just using a fork/join pool instead of a "simple" thread pool.

Also, this method should also react to the virtual threads config by using a thread factory that creates virtual threads if that flag is set.

We also probably want to set the `async` mode on the pool to `true`, rather than the default of `false`.  The default tends to not run tasks until the result is requested or polled, which increases latency.

File: server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+    private static class AsyncWriterThreadFactory implements ThreadFactory {
+        private final AtomicInteger threadNumber = new AtomicInteger(1);
+
+        @Override
+        public Thread newThread(@NonNull final Runnable r) {
+            return new Thread(r, "async-writer-" + threadNumber.getAndIncrement());
+        }
+    }
+}
```
Comment (jsync-swirlds) [inline server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java:145]: It's not clear what we need this for.  The default thread factory should work well (unless we want virtual threads).

**Review by jsync-swirlds — COMMENTED** [review]
A few items, mostly relating to the fork/join option.

File: server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java (hunk 3 above)
Comment (ata-nas) [reply server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java:145]: I do see some benefit in having a custom thread name, it would make it easier to look at a thread dump in order to see which thread is running a writer.

File: server/src/test/java/com/hedera/block/server/persistence/storage/PersistenceStorageConfigTest.java [hunk 4]
```diff
@@ -76,7 +77,12 @@ void testPersistenceStorageConfigStorageTypes(final StorageType storageType) {
                 CompressionType.NONE,
                 DEFAULT_COMPRESSION_LEVEL,
                 DEFAULT_ARCHIVE_ENABLED,
-                DEFAULT_ARCHIVE_BATCH_SIZE);
+                DEFAULT_ARCHIVE_BATCH_SIZE,
+                PersistenceStorageConfig.ExecutorType.THREAD_POOL,
+                6,
+                60,
+                false,
+                2048);
```
Comment (ata-nas) [inline server/src/test/java/com/hedera/block/server/persistence/storage/PersistenceStorageConfigTest.java:85]: I would suggest that you save these values in a field that would be easily reused. The values should be the config's default values. Reasoning is that first of all, the values are reused. A change in the values that would break the tests with the current value set will only have to be done in a single place (if requirements change so that an assert will no longer pass, the tests need to be adapted).

Consider for all places.

File: server/src/test/java/com/hedera/block/server/persistence/storage/PersistenceStorageConfigTest.java [hunk 5]
```diff
@@ -169,10 +190,132 @@ void testPersistenceStorageConfigCompressionTypes(final CompressionType compress
...
+                DEFAULT_ARCHIVE_ENABLED,
+                DEFAULT_ARCHIVE_BATCH_SIZE,
+                PersistenceStorageConfig.ExecutorType.THREAD_POOL,
+                6,
+                60,
+                false,
+                queueLimit);
+        assertThat(actual).returns(queueLimit, from(PersistenceStorageConfig::executionQueueLimit));
+    }
+
```
Comment (ata-nas) [inline server/src/test/java/com/hedera/block/server/persistence/storage/PersistenceStorageConfigTest.java:318]: It would be a good idea to parameterize these tests and to supply additional values as well as edge cases. See other tests in the same class for simple examples. The more cases we test the better.

File: server/src/test/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactoryTest.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        final PersistenceStorageConfig config = createConfig(ExecutorType.THREAD_POOL, true, 4, 60L, 100);
+
+        // When
+        final Executor executor = AsyncWriterExecutorFactory.createExecutor(config);
+
+        // Then
+        assertNotNull(executor);
+        System.out.println(executor.getClass().getName());
+        // Virtual thread executor invokes underneath ThreadPerTaskExecutor with specific factory for virtual
+        assertTrue(executor.getClass().getName().contains("ThreadPerTaskExecutor"));
```
Comment (ata-nas) [inline server/src/test/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactoryTest.java:74]: I am not sure what this test asserts? Judging by the name, it asserts the creation of an executor with type THREAD_POOL? If that is so, the above test already covers that with the enum source parameterized test.

File: server/src/test/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactoryTest.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+                PersistenceStorageConfig.StorageType.BLOCK_AS_LOCAL_FILE,
+                PersistenceStorageConfig.CompressionType.NONE,
+                3, // Default compression level
+                true, // Default archive enabled
+                1000, // Default archive batch size
+                executorType,
+                threadCount,
+                threadKeepAliveTime,
+                useVirtualThreads,
+                executionQueueLimit);
```
Comment (ata-nas) [inline server/src/test/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactoryTest.java:126]: @jsync-swirlds mentioned something about test dependency for creating configurations for tests, coming from services config dependency. For a future cleanup, we should take a look at that and use that on our side if applicable. For now, we have the test config builder which we could utilize:

```java
    // a sample code snippet that would help give you the configs you need before each test
    @BeforeEach
    void setUp() throws IOException {
        final Configuration config = new TestConfigBuilder(PersistenceStorageConfig.class)
                .withValue(PERSISTENCE_STORAGE_COMPRESSION_TYPE, "NONE")
                .withValue(PERSISTENCE_STORAGE_ARCHIVE_BATCH_SIZE, String.valueOf(ARCHIVE_GROUP_SIZE))
                .withValue(PERSISTENCE_STORAGE_LIVE_ROOT_PATH_KEY, testTempDir.resolve("live"))
                .withValue(PERSISTENCE_STORAGE_ARCHIVE_ROOT_PATH_KEY, testTempDir.resolve("archive"))
                .getOrCreateConfig();
        persistenceStorageConfig = config.getConfigData(PersistenceStorageConfig.class);
    }
```

You can consider to using this because it is very easy to setup with the specific values you actually need, the rest will be resolved with default values automatically, thus making your setup a lot easier and shorter.

**Review by ata-nas — COMMENTED** [review]
Good job @georgi-l95. I leave some nits, suggestions for cleanup and some general thoughts.

File: server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java (hunk 3 above)
Comment (georgi-l95) [reply server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java:145]: I added it for the exact reason @ata-nas mentioned. But I will remove it, if the team decides 👍

Comment (jsync-swirlds) [reply server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java:145]: It's not a _bad_ item; just more maintenance for a very small possible use case.

If we keep this, it _should_ follow the virtual threads setting, however.

File: server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+                TRACE,
+                "Creating thread pool executor with %d threads and queue limit of %d"
+                        .formatted(threadCount, queueLimit));
+
+        return new ThreadPoolExecutor(
+                threadCount,
+                threadCount,
+                threadKeepAliveTime,
+                TimeUnit.SECONDS,
+                new LinkedBlockingDeque<>(queueLimit),
```
[the remark this answers is not in the document]
Comment (jsync-swirlds) [reply server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java]: In most cases, linked is a more efficient choice here.
Having to rewrite arrays generally costs more than the slight memory overhead.

File: server/src/test/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactoryTest.java (hunk 7 above)
Comment (jsync-swirlds) [reply server/src/test/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactoryTest.java:126]: The [latest upgrade](https://github.com/hiero-ledger/hiero-block-node/pull/761) to the consensus libraries has already replaced the `TestConfigBuilder` with `ConfigurationBuilder`, which works more-or-less the same.

File: server/src/test/java/com/hedera/block/server/persistence/PersistenceInjectionModuleTest.java [hunk 9]
```diff
@@ -73,6 +75,12 @@ class PersistenceInjectionModuleTest {
     @Mock
     private Executor executorMock;
 
+    @Mock
+    private BlockNodeContext blockNodeContextMock;
+
+    @Mock
+    private Configuration configurationMock;
+
```
[the remark this answers is not in the document]
Comment (jsync-swirlds) [reply server/src/test/java/com/hedera/block/server/persistence/PersistenceInjectionModuleTest.java:83]: I definitely support dropping as many mocks as possible.
The less we mock, the more accurate our test results will be.

File: server/src/test/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactoryTest.java (hunk 7 above)
Comment (AlfredoG87) [reply server/src/test/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactoryTest.java:126]: Yes, if you rebase to the latest `main` head you will no longer be able to use the `TestConfigBuilder`

**Review by AlfredoG87 — COMMENTED** [review]
Looking good 👍

File: server/src/test/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactoryTest.java (hunk 7 above)
Comment (ata-nas) [reply server/src/test/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactoryTest.java:126]: Yes, and thanks @AlfredoG87 for introducing the change on our side!

File: server/src/main/java/com/hedera/block/server/persistence/storage/PersistenceStorageConfig.java (hunk 1 above)
Comment (georgi-l95) [reply server/src/main/java/com/hedera/block/server/persistence/storage/PersistenceStorageConfig.java]: Yes, you are right. I've changed the implementation to use dedicated pool.

File: server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java (hunk 2 above)
Comment (georgi-l95) [reply server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java]: Thank you for catching this. Changed the implementation to use dedicated pool. 👍

File: server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java (hunk 3 above)
Comment (georgi-l95) [reply server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java]: Yes, you're right. Added functionality to respect the virtual threads setting.

File: server/src/main/java/com/hedera/block/server/persistence/storage/PersistenceStorageConfig.java [hunk 10]
```diff
@@ -30,7 +30,12 @@ public record PersistenceStorageConfig(
         @Loggable @ConfigProperty(defaultValue = "ZSTD") CompressionType compression,
         @Loggable @ConfigProperty(defaultValue = "3") @Min(0) @Max(20) int compressionLevel,
         @Loggable @ConfigProperty(defaultValue = "true") boolean archiveEnabled,
-        @Loggable @ConfigProperty(defaultValue = "1_000") int archiveGroupSize) {
+        @Loggable @ConfigProperty(defaultValue = "1_000") int archiveGroupSize,
+        @Loggable @ConfigProperty(defaultValue = "THREAD_POOL") ExecutorType executorType,
+        @Loggable @ConfigProperty(defaultValue = "6") @Min(1) @Max(16) int threadCount,
+        @Loggable @ConfigProperty(defaultValue = "60") @Min(0) long threadKeepAliveTime,
+        @Loggable @ConfigProperty(defaultValue = "false") boolean useVirtualThreads,
+        @Loggable @ConfigProperty(defaultValue = "1024") @Min(64) @Max(2048) int executionQueueLimit) {
```
[the remark this answers is not in the document]
Comment (georgi-l95) [reply server/src/main/java/com/hedera/block/server/persistence/storage/PersistenceStorageConfig.java]: Added preconditions, thank you 👍

File: server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+     *
+     * @param config the persistence storage configuration containing executor settings
+     * @return an executor configured according to the settings
+     * @throws NullPointerException if config is null
+     */
+    @NonNull
+    public static Executor createExecutor(@NonNull final PersistenceStorageConfig config) {
+        requireNonNull(config);
+
+        LOGGER.log(TRACE, "Creating async writer executor of type: %s".formatted(config.executorType()));
```
[the remark this answers is not in the document]
Comment (georgi-l95) [reply server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java]: Changed to this formating, now that it's fixed 👍

File: server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        LOGGER.log(
+                TRACE,
+                "Creating thread pool executor with %d threads and queue limit of %d"
+                        .formatted(threadCount, queueLimit));
+
+        return new ThreadPoolExecutor(
+                threadCount,
+                threadCount,
+                threadKeepAliveTime,
+                TimeUnit.SECONDS,
```
[the remark this answers is not in the document]
Comment (georgi-l95) [reply server/src/main/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactory.java]: You're right, changed.

File: server/src/test/java/com/hedera/block/server/config/ServerMappedConfigSourceInitializerTest.java [hunk 13]
```diff
@@ -59,6 +59,11 @@ class ServerMappedConfigSourceInitializerTest {
         new ConfigMapping("persistence.storage.type", "PERSISTENCE_STORAGE_TYPE"),
         new ConfigMapping("persistence.storage.archiveEnabled", "PERSISTENCE_STORAGE_ARCHIVE_ENABLED"),
         new ConfigMapping("persistence.storage.archiveGroupSize", "PERSISTENCE_STORAGE_ARCHIVE_BATCH_SIZE"),
+        new ConfigMapping("persistence.storage.executorType", "PERSISTENCE_STORAGE_EXECUTOR_TYPE"),
+        new ConfigMapping("persistence.storage.threadCount", "PERSISTENCE_STORAGE_THREAD_COUNT"),
+        new ConfigMapping("persistence.storage.threadKeepAliveTime", "PERSISTENCE_STORAGE_THREAD_KEEP_ALIVE_TIME"),
+        new ConfigMapping("persistence.storage.useVirtualThreads", "PERSISTENCE_STORAGE_USE_VIRTUAL_THREADS"),
+        new ConfigMapping("persistence.storage.executionQueueLimit", "PERSISTENCE_STORAGE_EXECUTION_QUEUE_LIMIT"),
```
[the remark this answers is not in the document]
Comment (georgi-l95) [reply server/src/test/java/com/hedera/block/server/config/ServerMappedConfigSourceInitializerTest.java]: Good idea, done.

File: server/src/test/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactoryTest.java (hunk 6 above)
Comment (georgi-l95) [reply server/src/test/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactoryTest.java]: You're right, modified it to use the api you've mentioned.

File: server/src/test/java/com/hedera/block/server/persistence/storage/PersistenceStorageConfigTest.java (hunk 5 above)
Comment (georgi-l95) [reply server/src/test/java/com/hedera/block/server/persistence/storage/PersistenceStorageConfigTest.java]: True, done.

File: server/src/test/java/com/hedera/block/server/persistence/PersistenceInjectionModuleTest.java [hunk 14]
```diff
@@ -204,4 +212,33 @@ void testProvidesStreamValidatorBuilder() throws IOException {
...
+                PersistenceInjectionModule.providesBlockNodeEventHandler(
+                        subscriptionHandlerMock,
+                        notifierMock,
+                        blockNodeContextMock,
+                        serviceStatusMock,
+                        ackHandlerMock,
+                        asyncBlockWriterFactoryMock);
+
+        // Then
+        assertNotNull(handler);
```
[the remark this answers is not in the document]
Comment (georgi-l95) [reply server/src/test/java/com/hedera/block/server/persistence/PersistenceInjectionModuleTest.java]: You can check again, a lot have changed after the rebase 😆

File: server/src/test/java/com/hedera/block/server/persistence/storage/PersistenceStorageConfigTest.java (hunk 4 above)
Comment (georgi-l95) [reply server/src/test/java/com/hedera/block/server/persistence/storage/PersistenceStorageConfigTest.java]: Yep, done. Thank you

File: server/src/test/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactoryTest.java (hunk 7 above)
Comment (georgi-l95) [reply server/src/test/java/com/hedera/block/server/persistence/storage/write/AsyncWriterExecutorFactoryTest.java]: Thanks all, done.

**Review by ata-nas — APPROVED** [review]
LGTM. 🔥

**Review by jsync-swirlds — APPROVED** [review]

**Review by ata-nas — APPROVED** [review]
Good job! :fire:
